### PR TITLE
BUGFIX: Diffractometer responds to energy PV updates

### LINKED ISSUE #128: units & offset updates not correct in Diffractometer
Example was revised and tested.  The tests have exposed some problems with how changes in units and offset are handled by the `Diffractometer()` class.

Need to revise how `Diffractometer()` responds to changes in units and offset signals.  One problem observed is how the energy signal is affected when offset is changed.

_Originally posted by @prjemian in https://github.com/bluesky/hklpy/issues/121#issuecomment-824388061_

### PR DESCRIPTION
**Pull request by prjemian**
fix #128

### CODE REVIEW

File: hkl/tests/test_diffract.py [hunk 1]
```diff
@@ -127,14 +128,14 @@ def test_energy_units(fourc):
     # issue #86
     # changing units or offset changes .energy, not .calc.energy
     fourc.energy_units.put("eV")
-    assert fourc.calc.energy == 8.015
-    assert round(fourc.energy.get(), 1) == 8015
+    assert round(fourc.calc.energy, 6) == 8.015e-3
```
Comment (mrakitin) [inline hkl/tests/test_diffract.py:131]: Why did it change 6 orders of magnitude?

**Review by mrakitin — COMMENTED** [review]
I have a question about the test, otherwise looks good.

File: hkl/tests/test_diffract.py (hunk 1 above)
Comment (prjemian) [reply hkl/tests/test_diffract.py:131]: What happened is that the units were changed from keV to eV which caused fourc.energy (8.015 eV) to be converted in fourc.calc.energy (8.015e-3 eV).  This now tests the interim state that happens as one changes both eV and value.

Comment (prjemian) [reply hkl/tests/test_diffract.py:131]: Tests might be cleaner if they were separated.  As it stands now, later tests rely on the state from previous tests.

File: hkl/tests/test_diffract.py [hunk 2]
```diff
@@ -116,26 +118,37 @@ def test_energy_units(fourc):
...
     assert round(fourc.energy.get(), 6) == 8
 
+
+def test_energy_units_issue86(fourc):
     # issue #86
     # changing units or offset changes .energy, not .calc.energy
+    fourc.energy.put(8)
+    fourc.energy_offset.put(0.015)
     fourc.energy_units.put("eV")
+    # test interim state when fourc.energy value has not change but units have
```
Comment (mrakitin) [inline hkl/tests/test_diffract.py:143]:
```suggestion
    # test interim state when fourc.energy value has not changed but units have
```

File: hkl/tests/test_diffract.py (hunk 1 above)
Comment (mrakitin) [reply hkl/tests/test_diffract.py:144]: So, the `fourc.calc.energy` is always measured in `keV` units and that cannot be changed, right?

**Review by mrakitin — APPROVED** [review]
Thanks for separating the tests, they are more readable/understandable now.

I left a question to confirm and a minor comment correction, otherwise looks good.

**Comment by prjemian** [comment]
@mrakitin: Yes, that's right.  The engineering units in the `hkl.calc` module are set by this code to keV (energy) and Angstrom (wavelength): https://github.com/bluesky/hklpy/blob/9b5c6c8cd602da122dd890457445ba903c682969/hkl/calc.py#L170-L186

Since the units of wavelength and crystal lattice unit cell length must be able to cancel, this imposes the additional rule that crystal lattice unit cell lengths must also be in angstrom, the units of `hkl.calc.wavelength`.

**Pull request merged**